esp-hal/src/gpio/mod.rs

```diff
@@ -645,19 +645,34 @@ pub fn connect_high_to_peripheral(signal: InputSignal) {
 }
 
 #[doc(hidden)]
-pub trait PinType {}
+pub trait BooleanType {
+    const VALUE: bool;
+}
 
 #[doc(hidden)]
-pub trait IsOutputPin: PinType {}
+pub struct True {}
 
-#[doc(hidden)]
-pub trait IsInputPin: PinType {}
+impl BooleanType for True {
+    const VALUE: bool = true;
+}
 
 #[doc(hidden)]
-pub trait IsAnalogPin: PinType {}
+pub struct False {}
+
+impl BooleanType for False {
+    const VALUE: bool = false;
+}
 
 #[doc(hidden)]
-pub trait IsTouchPin: PinType {}
+pub trait PinType {
+    type IsOutput: BooleanType;
```

Comment:
I like this idea

Reply:
I'm a bit worried if this is just more complicated than the marker traits, for no good reason.